website/server/libs/groupTasks.js

```diff
@@ -1,61 +1,268 @@
+        masterEntry => masterEntry.date > historyEntry.date,
+      );
+      if (historyIndex === -1) {
+        masterTask.history.push(historyEntry);
+      } else {
+        masterTask.history.splice(historyIndex, 0, historyEntry);
+      }
+    } else if (!groupMemberTask.completed && !approvalRequired) {
+      // User uncompleted a task which did not need approval
+      // Loop backwards through the history and splice out the user's last entry
+      if (masterTask.history && masterTask.history.length > 0) {
+        for (let i = masterTask.history.length - 1; i >= 0; i -= 1) {
+          if (masterTask.history[i].userId === groupMemberTask.userId) {
+            masterTask.history.splice(i, 1);
+            break;
+          }
+        }
+      }
+    }
+
+    promises.push(masterTask.save());
+    await Promise.all(promises);
+  }
 }
 
-async function _evaluateAllAssignedCompletion (masterTask) {
+async function _evaluateAllAssignedCompletion (masterTask, groupMemberTask) {
   let completions;
   if (masterTask.group.approval && masterTask.group.approval.required) {
     completions = await Tasks.Task.count({
       'group.taskId': masterTask._id,
       'group.approval.approved': true,
     }).exec();
-    completions += 1;
+    // Since an approval is not yet saved into the group member task, count it
+    // But do not recount a group member completing an already approved task
+    if (!groupMemberTask.completed) completions += 1;
   } else {
     completions = await Tasks.Task.count({
       'group.taskId': masterTask._id,
       completed: true,
     }).exec();
   }
-  if (completions >= masterTask.group.assignedUsers.length) {
-    await _completeMasterTask(masterTask);
+  await _completeOrUncompleteMasterTask(
+    masterTask,
+    completions >= masterTask.group.assignedUsers.length,
+  );
+}
+
+async function groupTaskCompleted (groupMemberTask, user, now) {
+  const masterTask = await Tasks.Task.findOne({
+    _id: groupMemberTask.group.taskId,
+  }).exec();
+
+  if (!masterTask || !masterTask.group) return false;
+
+  if (masterTask.history && masterTask.history.length > 0) {
+    // I hope we don't really need to loop the history and check every date
+    // (User with a late day start in Australia vs and early one in Hawaii?)
+    const taskLastHistory = masterTask.history[masterTask.history.length - 1];
+    // Check if the history entry has a user who completed the task
+    // TODO When group scoring is implemented, add history entries for group missed dailies
+    if (taskLastHistory.userId) {
+      const lastCompletingUser = await Users.findById(taskLastHistory.userId, 'preferences');
+      if (lastCompletingUser) {
+        // Check what the completing user's "day" was when the task was completed
+        const taskLastCompletedDay = startOfDay(
+          defaults(
+            { now: moment(taskLastHistory.date) },
+            lastCompletingUser.preferences.toObject(),
```

Comment:
is toObject necessary? I'm seeing it's called here but for example not in scoreTask